src/NumberTheory/vinberg.jl

```diff
@@ -0,0 +1,336 @@
+  l = nrows(Q)
+  S = snf(Q)
+  possible_root_lengths = divisors(2 * S[l, l])
+  if isempty(root_lengths)
+    return possible_root_lengths
+  end
+  result = ZZRingElem[]
+  for r in root_lengths
+    if abs(r) in possible_root_lengths
+      push!(result, r)
+    else
+      @vprintln :Vinberg 1 "$r is not a possible root length"
+    end
+  end
+  @req !isempty(result) "No possible root lengths found"
+  return result
+end
+
+@doc raw"""
+    _distance_indices(upper_bound, root_lengths::Vector{ZZRingElem}) -> Vector{Tuple{Int, ZZRingElem}}
+
+Returns all possible tuples $(n, l)$ with `n` a natural number and `l` contained in `root_lengths`,
+sorted by increase of the value $\frac{n^2}{l}$, stopping by `upper_bound`.
+"""
+function _distance_indices(upper_bound, root_lengths::Vector{ZZRingElem})
+
+  result = Tuple{Int,ZZRingElem}[]
+
+  for l in root_lengths
+    # consider for the `upper_bound` $u$ $\frac{n^2}{l} < u \implies n < \sqrt{l*u}$
+    x = isqrt(abs(l * upper_bound))
+    for n in 1:x
+      push!(result, (n, l))
+    end
+  end
+
+  sort!(result; by=x -> (x[1]^2) // abs(x[2]))
+  return result
+end
+
+@doc raw"""
+    _rescale_primitive(v::ZZMatrix) -> ZZMatrix
+
+Check if `v` is primitive, which is equivalent to checking whether the greatest common divisor `g` of the entries of `v`
+is equal to 1. If `v` is not primitive, we divide `v` by `g`.
+"""
+function _rescale_primitive(v::ZZMatrix)
+  g = reduce(gcd, v)
+  return isone(g) ? v : v / g
+end
+
+@doc raw"""
+    _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+
+Check if the reflection by `v` preserves the lattice, i.e. check if for `v` a row vector
+$\frac{2.v.Q}{v^2}$ is an integer matrix. 
+"""
+function _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+  A = Q * transpose(v)
+  b = (v * A)[1, 1]
+  return all(iszero(mod(2*x, b)) for x in A)
+end
+  
+@doc raw"""
+    _crystallographic_condition(Q::ZZMatrix, v::ZZMatrix)
+
+Check if the reflection by `v` preserves the lattice, i.e. check if for `v` a row vector
+$\frac{2 v Q}{vQv^t}$ is an integer matrix
+where `k` is $vQv^t$.
+"""
+@inline function _crystallographic_condition(Qv::ZZMatrix, k::ZZRingElem)
+  return all(iszero(mod(2*x, k)) for x in Qv)
+end
+
+@doc raw"""
+    _check_coorientation(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix, v0::ZZMatrix)
+
+First check whether `v` has non-obtuse angles with all roots `r` already found, by checking that $v.r \geq 0$.
+"""
+function _has_non_obtuse_angles(Q::ZZMatrix, roots::Vector{ZZMatrix}, v::ZZMatrix)
+  Qv = Q * transpose(v)
+  for r in roots
+    if (r * Qv)[1, 1] < 0
+      return false
+    end
+  end
+  return true
+end
+
+# non-allocating version
+function _has_non_obtuse_angles!(tmp::ZZMatrix, Qv::ZZMatrix, roots::Vector{ZZMatrix})
+  for r in roots
+    #tmp = r*Qv
+    mul!(tmp, r, Qv)
+    if !is_zero_entry(tmp,1, 1) && !is_positive_entry(tmp, 1, 1)
+      return false
+    end
+  end
+  return true
+end
```

Comment:
@ElMerkel  
This is the non-allocating version I was referring to. Note however that using `mul!` is unsafe and can lead to segmentation faults if it gets bad input, e.g. matrices of a wrong size.  
Hence it is only ever worth using in time critical places.

Reply:
Note that you can still make it perfectly safe by adding an input validation check yourself, i.e. in this case: that the number of rows resp. columns of the two matrices fit together.